### PR TITLE
Change transformation node size on column selected (fixes #28)

### LINKED ISSUE #28: When multiple columns have to be shown for the same transformation, they are all shown in the same position

### PR DESCRIPTION
**Pull request by eze210**

### CODE REVIEW

File: spark-board-ui/src/App.jsx [hunk 1]
```diff
@@ -97,10 +106,8 @@ export default function App() {
                 return node;
             }));
         }
-    }, [settings, nodes, setNodes]);
-
-    // controls the column for which the column graph is shown
-    const [selectedColumn, setSelectedColumn] = useColumnGraphState(null);
+        setSelectedColumn(selectedColumn);
+    }, [settings, nodes, setNodes, setSelectedColumn]);
```
Comment (alijdens) [inline spark-board-ui/src/App.jsx:110]: I think it should depend on `selectedColumn` and not `setSelectedColumn`

File: spark-board-ui/src/column.jsx [hunk 2]
```diff
@@ -7,14 +7,14 @@ import { Handle, Position } from 'reactflow';
...
-            <p>{ data.name }</p>
-            <Handle type="source" position={Position.Right} id="source" />
-        </div>
+        <>
+            <div className="column-node__container" >
+                <Handle type="target" position={Position.Left} id="target" />
+                <p>{ data.name }</p>
+                <Handle type="source" position={Position.Right} id="source" />
+            </div>
+        </>
```
Comment (alijdens) [inline spark-board-ui/src/column.jsx:17]: I don't see why this change is necessary

File: spark-board-ui/src/App.jsx [hunk 3]
```diff
@@ -78,7 +75,22 @@ export default function App() {
             // orgnaize the nodes
             organizeNodes();
         }
-    }, [nodesInitialized]);
+    }, [nodesInitialized, settings]);
```
Comment (alijdens) [inline spark-board-ui/src/App.jsx:78]: Won't this make the nodes reorder when the settings are updated?

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js [hunk 4]
```diff
@@ -71,7 +71,7 @@ export function useDagAnimation(nodes, edges, targetPositions, setNodes) {
 
         // wait for the next frame
         animationId.current = requestAnimationFrame(onFrame)
-    }
+    }, [state, nodeIdMapping, nodes, setNodes]);
```
Comment (alijdens) [inline spark-board-ui/src/useDagAnimation/useDagAnimation.js:74]: This function is calling `setNodes` which change the `nodes`, so it's a circular dependency

File: spark-board-ui/src/App.jsx (hunk 1 above)
Comment (eze210) [reply spark-board-ui/src/App.jsx]: Now it depends on the columnGraph, so the dependency on selectedColumn isn't needed

File: spark-board-ui/src/column.jsx (hunk 2 above)
Comment (eze210) [reply spark-board-ui/src/column.jsx]: Reverted

File: spark-board-ui/src/App.jsx (hunk 3 above)
Comment (eze210) [reply spark-board-ui/src/App.jsx:78]: Yes, I liked it when testing, but I will revert it

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js (hunk 4 above)
Comment (eze210) [reply spark-board-ui/src/useDagAnimation/useDagAnimation.js]: You're right, reverting this part of the change

File: spark-board-ui/src/App.jsx (hunk 1 above)
Comment (eze210) [reply spark-board-ui/src/App.jsx]: Removed that setter

File: spark-board-ui/src/App.jsx [hunk 5]
```diff
@@ -48,11 +54,14 @@ export default function App() {
         column: ColumnNode
     }), []);
 
-    const [transformationNodePositions, setTransformationNodePositions] = React.useState(new Map(
+    const [transformationNodeTargetPositions, setTransformationNodeTargetPositions] = React.useState(new Map(
```
Comment (alijdens) [inline spark-board-ui/src/App.jsx:57]:
```suggestion
    // positions where the nodes are being pulled to (if the animation is enabled)
    const [transformationNodeTargetPositions, setTransformationNodeTargetPositions] = React.useState(new Map(
```

File: spark-board-ui/src/App.jsx [hunk 6]
```diff
@@ -48,11 +54,14 @@ export default function App() {
...
         model_initialNodes.map(node => [node.id, node.position])
     ));
+    const [transformationNodeCurrentPositions, setTransformationNodeCurrentPositions] = React.useState(new Map(
```
Comment (alijdens) [inline spark-board-ui/src/App.jsx:60]:
```suggestion
    // positions where the nodes are drawn (when nodes are being animated)
    const [transformationNodeCurrentPositions, setTransformationNodeCurrentPositions] = React.useState(new Map(
```

File: spark-board-ui/src/App.jsx [hunk 7]
```diff
@@ -66,20 +75,35 @@ export default function App() {
...
     }, [nodesInitialized, settings]);
 
+    useEffect(() => {
+        setNodes(nodes.map(node => {
+            if (node.type == "transformation") {
+                node.position = transformationNodeCurrentPositions.get(node.id);
+            }
+            return node;
+        }));
+    }, [transformationNodeCurrentPositions]);
```
Comment (alijdens) [inline spark-board-ui/src/App.jsx:85]: I don't understand why this would fix the problem. What's the difference between calling `setNodes` here instead of the `useDagAnimation` hook?

File: spark-board-ui/src/App.jsx [hunk 8]
```diff
@@ -35,7 +35,13 @@ export default function App() {
             selectedTransformation !== null && selectedTransformation.data == node.data
         );
 
-        node.position = (node.type == "transformation" ? {x: 0, y: 0} : {x: 0, y: 25});
+        if (node.position === undefined) {
```
[the remark this answers is not in the document]
Comment (eze210) [reply spark-board-ui/src/App.jsx]: It shouldn't be. Reverted

File: spark-board-ui/src/App.jsx (hunk 7 above)
Comment (eze210) [reply spark-board-ui/src/App.jsx]: The problem is that the old `setNodes` was using the nodes state with the value that was received in the `useDagAnimation` parameter, so we had a race-condition when the onFrame was executed in parallel to the effect that updates the node sizes.
Making the `onFrame` to set the new positions' state and the new effect to update the nodes makes that the two effects aren't executed in parallel.
Another solution would be to pass a lambda expression to the `setNodes` inside `onFrame`, but I liked more this solution

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js [hunk 9]
```diff
@@ -13,12 +13,17 @@ import { stepSimulation } from './physics';
  * @param {Array} nodes react-flow nodes.
  * @param {Array} edges react-flow edges.
  * @param {Map} targetPositions Map from the node IDs to the target positions (objects with `x` and `y`).
- * @param {Callback} setNodes react-flow state setter to update the nodes.
+ * @param {Callback} setCurrentPositions react-flow state setter to update the node positions.
```
Comment (alijdens) [inline spark-board-ui/src/useDagAnimation/useDagAnimation.js:16]: ?

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js [hunk 10]
```diff
@@ -13,12 +13,17 @@ import { stepSimulation } from './physics';
...
  * @returns Two elements:
  *          1. Callback that will start the animation when called.
  *          2. Callback that will update the node positions with the ones given.
  */
 export function useDagAnimation(nodes, edges, targetPositions, setNodes) {
+    // positions where the nodes are being pulled to (if the animation is enabled)
+    const [currentPositions, setCurrentPositions] = useState(new Map(
+        nodes.map(node => [node.id, node.position])
+    ));
+
```
Comment (alijdens) [inline spark-board-ui/src/useDagAnimation/useDagAnimation.js:26]: do you think it makes sense to make this a state?

it's not being shared anywhere outside of this hook and only updated internally. It feels like it's an unnecesary middleman.

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js [hunk 11]
```diff
@@ -54,25 +59,32 @@ export function useDagAnimation(nodes, edges, targetPositions, setNodes) {
...
+            })));
         }
 
         // wait for the next frame
         animationId.current = requestAnimationFrame(onFrame)
     }
 
+    // update the transformation nodes when the current position changes
+    useEffect(() => {
+        setNodes((nds) => nds.map(node => {
```
Comment (alijdens) [inline spark-board-ui/src/useDagAnimation/useDagAnimation.js:80]: the real fix was to use a callback inside of `setNodes`, right?

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js (hunk 10 above)
Comment (eze210) [reply spark-board-ui/src/useDagAnimation/useDagAnimation.js:26]: I liked this because the effect would be the "node updater" and, if we add another state, we could just add a new dependency to the effect... But yes, maybe it's useless right now and it would be a good change in case we add that hypotetical new state in the future

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js (hunk 11 above)
Comment (eze210) [reply spark-board-ui/src/useDagAnimation/useDagAnimation.js:80]: The fix is not having a race condition XD

File: spark-board-ui/src/useDagAnimation/useDagAnimation.js (hunk 9 above)
Comment (eze210) [reply spark-board-ui/src/useDagAnimation/useDagAnimation.js:16]: Forgot to revert this part

**Review by alijdens — APPROVED** [review]
LGTM
